tiny_dnn/core/framework/tensor.h

```diff
@@ -26,83 +25,67 @@ namespace tiny_dnn {
- * the rightmost index “varies the fastest”.
  *
- * Data is held by a std::vector with 64 bytes alignment.
- * Unmutable if kConst == true
+ * U is type of the data stored
+ * Storage is type of underlying storage
  */
-template <typename U         = float_t,
-          size_t kDimensions = 4,
-          bool kConst        = false,
-          typename Allocator = aligned_allocator<U, 64>>
+template <typename U = float_t, typename Storage = xt::xarray<U>>
 class Tensor {
-  // Define constant types for constant Tensor,
-  // and mutable ones for mutable Tensor
-  typedef typename std::conditional<kConst,
-                                    const TensorStorage<U, Allocator>,
-                                    TensorStorage<U, Allocator>>::type
-    TensorStorageType;
-  typedef typename std::conditional<kConst, const U *, U *>::type UPtr;
-  typedef typename std::shared_ptr<TensorStorageType> TensorStoragePointer;
-  typedef typename std::conditional<
-    kConst,
-    typename std::vector<U, Allocator>::const_iterator,
-    typename std::vector<U, Allocator>::iterator>::type StorageIterator;
+  typedef U *UPtr;
 
  public:
   /**
    * Initializes an empty tensor.
    * @return
    */
-  Tensor() {
-    offset_ = size_ = size_t(0);
-    storage_ptr_    = std::make_shared<TensorStorageType>();
-  }
+  Tensor() { storage_ = xt::xarray<U>(); }
 
   /**
-   * Constructor that assepts an array of shape and create a Tensor with that
-   * shape. For example, given shape = {2,3,4,5,6}, tensor
-   * will be of size 2x3x4x5x6
+   * Constructor that assepts an initializer list of shape and create a
+   * Tensor with that shape. For example, given shape = {2,3,4,5,6}, tensor
+   * will be of size 2x3x4x5x6. Note: tensor isn't initialized by default
    * @param shape array containing N integers, sizes of dimensions
    * @return
    */
-  explicit Tensor(const std::array<size_t, kDimensions> &shape) {
-    offset_      = size_t(0);
-    shape_       = shape;
-    size_        = product(shape);
-    storage_ptr_ = std::make_shared<TensorStorageType>(shape);
+  explicit Tensor(std::vector<size_t> const &shape) {
+    storage_ = xt::xarray<U>(shape);
   }
 
   /**
-   * Constructor that assepts a vector of shape and create a Tensor with that
-   * shape. For example, given shape = {2,3,4,5,6}, tensor
-   * will be of size 2x3x4x5x6
+   * Constructor that assepts an initializer list of shape and create a
+   * Tensor with that shape. For example, given shape = {2,3,4,5,6}, tensor
+   * will be of size 2x3x4x5x6. Note: tensor isn't initialized by default
    * @param shape array containing N integers, sizes of dimensions
    * @return
    */
-  explicit Tensor(const std::vector<size_t> &shape) {
-    offset_ = size_t(0);
-    size_   = product(shape);
-    std::copy(shape.begin(), shape.end(), shape_.begin());
-    storage_ptr_ = std::make_shared<TensorStorageType>(shape);
+  explicit Tensor(std::initializer_list<size_t> const &shape) {
```

Comment:
I think this is also something to be a bit cautious about, since people might assume that the initializer list is used as the initial values of the Tensor instead of the shape.

In xtensor we have a static named constructor `xarray<double>::from_shape(std::initializer_list...)` for this purpose now.